src/plugins/compressors/sz_threadsafe_plugin.cc

```diff
@@ -0,0 +1,424 @@
+
+    size_t r[] = {
+     pressio_data_get_dimension(output, 0),
+     pressio_data_get_dimension(output, 1),
+     pressio_data_get_dimension(output, 2),
+     pressio_data_get_dimension(output, 3),
+     pressio_data_get_dimension(output, 4),
+    };
+    size_t ndims = pressio_data_num_dimensions(output);
+
+    int status = SZ_NSCS;
+    pressio_dtype type = pressio_data_dtype(output);
+    /*void* decompressed_data = SZ_decompress_customize(
+        app.c_str(),
+        user_params,
+        libpressio_type_to_sz_type(type),
+        (unsigned char*)pressio_data_ptr(input, nullptr),
+        pressio_data_get_dimension(input, 0),
+        r[4],
+        r[3],
+        r[2],
+        r[1],
+        r[0],
+        &status
+        );*/
+    void* decompressed_data=SZ_decompress_customize_threadsafe("SZ",&threadsafe_params,SZ_FLOAT,(unsigned char*)input->data(),ndims,
+		    r[4],r[3],r[2],r[1],r[0],&status);
+    
+    *output = pressio_data::move(type, decompressed_data, ndims, r, pressio_data_libc_free_fn, nullptr);
+    return 0;
+  }
+
+  int major_version() const override {
+    return SZ_VER_MAJOR;
+  }
+  int minor_version() const override {
+    return SZ_VER_MINOR;
+  }
+  int patch_version() const override {
+    return SZ_VER_BUILD;
+  }
+  int revision_version () const { 
+    return SZ_VER_REVISION;
+  }
+
+  const char* version() const override {
+    return sz_version.c_str(); 
+  }
+
+  const char* prefix() const override {
+    return "sz_threadsafe";
+  }
+
+  pressio_options get_metrics_results_impl() const override {
+    pressio_options sz_threadsafe_metrics;
+#if HAVE_WRITESTATS
+    set(sz_threadsafe_metrics, "sz_threadsafe:use_mean", sz_threadsafe_stat.use_mean);
+    set(sz_threadsafe_metrics, "sz_threadsafe:block_size", (unsigned int)sz_threadsafe_stat.blockSize);
+    set(sz_threadsafe_metrics, "sz_threadsafe:lorenzo_blocks", (unsigned int)sz_threadsafe_stat.lorenzoBlocks);
+    set(sz_threadsafe_metrics, "sz_threadsafe:regression_blocks", (unsigned int)sz_threadsafe_stat.regressionBlocks);
+    set(sz_threadsafe_metrics, "sz_threadsafe:total_blocks", (unsigned int)sz_threadsafe_stat.totalBlocks);
+    set(sz_threadsafe_metrics, "sz_threadsafe:huffman_tree_size", (unsigned int)sz_threadsafe_stat.huffmanTreeSize);
+    set(sz_threadsafe_metrics, "sz_threadsafe:huffman_coding_size", (unsigned int)sz_threadsafe_stat.huffmanCodingSize);
+    set(sz_threadsafe_metrics, "sz_threadsafe:huffman_node_count", (unsigned int)sz_threadsafe_stat.huffmanNodeCount);
+    set(sz_threadsafe_metrics, "sz_threadsafe:unpredict_count", (unsigned int)sz_threadsafe_stat.unpredictCount);
+
+    set(sz_threadsafe_metrics, "sz_threadsafe:lorenzo_percent", sz_threadsafe_stat.lorenzoPercent);
+    set(sz_threadsafe_metrics, "sz_threadsafe:regression_percent", sz_threadsafe_stat.lorenzoPercent);
+    set(sz_threadsafe_metrics, "sz_threadsafe:huffman_compression_ratio", sz_threadsafe_stat.huffmanCompressionRatio);
```

Comment:
This doesn't do what is indented.  We should probably  try to get Sheng to fix the write stat's implementation to be threadsafe, and if he won't, we need to remove this code in the threadsafe version since it will never be correct in multi-threaded uses.